### ISSUE TITLE
Writing to unique input parameter arrays in entry points causes errors in the generated C code

### ISSUE THREAD

**Issue by nqpz** [diku-dk/futhark#388]
This is related to the work-in-progress memory block optimisations.

Consider this `main` function:
```
let main (xs: *[#n]i32): [n]i32 =
  map (+ 1) xs
```
In both the CPU and GPU pipelines, this will result in a program with a single memory allocation, namely one for the map.  I have an optimisation that currently reasons that since `xs` is unique, the map can try reusing the memory of `xs` instead of requiring a new allocation.  After size and index analysis, the compiler will accept the reuse, and a subsequent simplifier will then remove the allocation statement.

Reusing unique parameter memory seems like a trick that is okay to do (assuming proper interference analysis etc. is also performed), since the caller will not use the memory again.  There are at least two cases:

  + In stand-alone generated programs, entry point parameter memory can always said to be "unique" even when the programmer has not manually declared them to be so, in that its input memory contains the parsed values and is not directly controllable by the user.  The optimisation will always be useful in this case, and it cannot confuse any user.
  + In generated libraries, reusing unique entry point parameter memory means that the programmer using the library must make sure not to reuse the memory sent to library -- but only if the iibrary programmer chooses to manually make input parameters unique.  I think this is a slippery slope.

So, the concept of unique parameters in *entry point* functions seems a bit murky.  I think the best approach is to only perform the optimisation for stand-alone programs, not libraries.  I have a few programs that benefit from this feature.  In any case, my problem is this:

When a generated program is run with `-r N` -- used by e.g. `futhark-bench` -- it performs both a warmup run, and, if N > 1, runs the entry point function more than once, which means that reusing input memory in one run can cause the next run to give a wrong result.  This is easy to fix by copying all unique memory blocks before each run.  This will give an input copying overhead, but it only relates to benchmarking, and the effect is a possibly reduced memory footprint.

Opinions?  Downside: Slightly more copying when benchmarking.  Upside: Reduced memory footprint in some standalone programs.

**Comment by Athas** [comment]
Yes, I have known about this for a while.  You don't actually need memory block merging to show this; the following program will produce spurious results when benchmarked:

```
let main (xs: []i32) =
  xs with [0] <- xs[0] + 1
```

I did not bother fixing it, since it did not occur in practise.  But now it does.

I don't think the situation is murky at all.  The right solution is to copy unique parameters when running standalone executables.  In libraries, it is fine to make destructive updates to the input - that's what the uniqueness types mean!

**Comment by nqpz** [comment]
Okay, I will have the benchmarking code copy its unique parameters before each run.

**Comment by nqpz** [comment]
Closed in ef351e2f7bf7bfc22a428c30ae909a30f9c08957.
